src/libsyntax/feature_gate.rs

```diff
@@ -682,36 +817,37 @@ pub fn check_for_pushpop_syntax(f: Option<&Features>, diag: &Handler, span: Span
 }
 
 struct Context<'a> {
-    features: Vec<&'static str>,
+    features: Features,
     span_handler: &'a Handler,
     cm: &'a CodeMap,
     plugin_attributes: &'a [(String, AttributeType)],
 }
 
-impl<'a> Context<'a> {
-    fn enable_feature(&mut self, feature: &'static str) {
-        debug!("enabling feature: {}", feature);
-        self.features.push(feature);
-    }
-
-    fn gate_feature(&self, feature: &str, span: Span, explain: &str) {
-        let has_feature = self.has_feature(feature);
-        debug!("gate_feature(feature = {:?}, span = {:?}); has? {}", feature, span, has_feature);
-        if !has_feature && !self.cm.span_allows_unstable(span) {
-            emit_feature_err(self.span_handler, feature, span, GateIssue::Language, explain);
+macro_rules! gate_feature_fn {
+    ($cx: expr, $has_feature: expr, $span: expr, $name: expr, $explain: expr) => {{
+        let (cx, has_feature, span, name, explain) = ($cx, $has_feature, $span, $name, $explain);
```

Comment:
@LeoTestard couldn't you make `gate_feature_fn` an ordinary function instead of macro? e.g.
```rust
fn feature_gate_fn<F>(cx: &Context, has_feature: F, span: Span, name: &str, explain: &str)
    where F: Fn(&Features) -> bool
{ ... }
```